Prose/ProseLib/Sources/ConversationFeature/Chat/Chat.swift

```diff
@@ -254,20 +262,64 @@ struct ChatView: NSViewRepresentable {
     signposter.endInterval(#function, interval)
   }
 
-  func updateMessages(to messages: [ProseCoreViewsMessage], in webView: WKWebView) {
-    logger.trace("Updating \(Self.self): \(messages.count, privacy: .public) messages")
-
+  func updateMessages(
+    to messages: [ProseCoreViewsMessage],
+    in webView: WKWebView,
+    coordinator: Coordinator
+  ) {
     let interval = signposter.beginInterval(#function, id: self.signpostID)
 
-    let jsonData: Data = try! JSONEncoder().encode(messages)
-    let json = String(data: jsonData, encoding: .utf8) ?? "[]"
+    let messages = IdentifiedArrayOf<ProseCoreViewsMessage>(uniqueElements: messages)
+    let oldMessages: IdentifiedArrayOf<ProseCoreViewsMessage> = coordinator.messages
+    defer { coordinator.messages = messages }
+
+    let ids = OrderedSet<Message.ID>(messages.map(\.id))
+    let oldIds: OrderedSet<Message.ID> = coordinator.messages.ids
+
+    let insertedIds: OrderedSet<Message.ID> = ids.subtracting(oldIds)
+    let removedIds: OrderedSet<Message.ID> = oldIds.subtracting(ids)
+    let updatedIds: [Message.ID] = ids.intersection(oldIds)
+      .filter { messages[id: $0] != oldMessages[id: $0] }
+
+    for messageId in removedIds {
+      self.rectractMessage(withId: messageId, in: webView)
```

Comment:
```suggestion
      self.retractMessage(withId: messageId, in: webView)
```

Reply:
For some reason, GitHub doesn't show me this one in the file explorer. I can't commit the other proposition at the same time from here. I will do this in a commit on my machine (it won't set the co-author).

Reply:
Thank you for the catch by the way!